### ISSUE TITLE
DigitalOcean Support

### ISSUE THREAD

**Issue by Immortalin** [linuxkit/linuxkit#3131]
It would be great to add support for digitalocean. It can be done pretty easily by ensuring the right files are in place like this: https://github.com/bontibon/digitalocean-alpine

**Comment by Immortalin** [comment]
@justincormack any chance to use code from @bontibon's project? I use it in production and it works pretty well as an alpine distro. I think it should be quite straightforward to add DO support to Linuxkit since it's also built on alpine

**Comment by justincormack** [comment]
The relevant parts are different enough that I am not sure you could reuse
the code. I dont think it would be too hard to do though.

**Comment by Immortalin** [comment]
Any advice where to start? I.e. how to implement the /dev mappings etc.

**Comment by Immortalin** [comment]
The scaleway example looks surprisingly sparse

**Comment by Sh4d1** [comment]
@Immortalin I'm not sure but you could try to do the same as for the Scaleway provider. To build an image on Scaleway it's just a snapshot from a volume. Basically it's just copy the iso on a "build" server, use `dd` to copy it on an emtpy drive, and make a snapshot then an image of this volume.

**Comment by Sh4d1** [comment]
But you'll have to check how exactly is a DO droplet booting, you may be running into some issues there

**Comment by douglasmiranda** [comment]
Digital Ocean is enabling the importing of custom images. Could you guys tell me if one of these examples is a good fit? https://github.com/linuxkit/linuxkit/tree/master/examples

**Comment by Sh4d1** [comment]
@douglasmiranda do you have more information about DO's custom images?

**Comment by douglasmiranda** [comment]
I'm waiting for a response from DO. I will come back with the details as soon as I know. =]

**Comment by douglasmiranda** [comment]
Ok, I have some info that may be useful:

**Are there limits to the type of images you can import?**
Yes, you can upload any linux OS that supports ext3/4 file systems and has cloudinit
0.7.5,cloudbase-init, coreos-cloudinit, ignition or bsd-cloudinit installed.

**Which type of files can I upload:**
Currently, raw, qcow, ,qcow2, vhd, vhdx, vdi and vmdk formats. Also accept compressed gzip,
and BZip2 compressions. Note that ISO files are not supported at this time, expect ISO support
towards the end of 2018.

**I can’t SSH into my Droplet started from a custom image, what should I do?**

Symptoms: When trying to SSH into a Droplet, you are receiving a prompt to enter a root
password even though you don’t have one.

A. This is a known issue with Images that have an older version of cloud init. The
quick fix is to update the image with cloudinit 0.7.5 and re-import the Image.

They said these images work: https://docs.openstack.org/image-guide/obtain-images.html

**Comment by justincormack** [comment]
We don't at present have a version of cloud-init for LinuxKit. The full spec is very complex, it might be worth seeing what the DO cloudinit config actually looks like and implement support just for that subset...

**Comment by SvenDowideit** [comment]
the coreos cloud init code (written in go) is pretty simple, and is a functional subset.
I ended up forking a copy into rancherOS last year to add a few fixes.

that code works well enough for DO, AWS, GCE, Azure, VMWare and i can't recal what else i used to test the heck out of :)

if you don't want to re-write that subset in rust >:} maybe you could work with Darren and the RancherOS crew?

**Comment by justincormack** [comment]
@SvenDowideit yes definitely it is the only sane version. Didnt realise rancher had a forked version...

**Comment by Immortalin** [comment]
https://blog.digitalocean.com/custom-images/

**Comment by gianarb** [comment]
This is how the API works https://developers.digitalocean.com/documentation/v2/#create-a-custom-image
The fact that they require a valid URL makes the upload more complicated?

**Comment by justincormack** [comment]
@gianarb thats what we already do for the Packet client which uses iPXE over http, it is slightly inconvenient but you can use `ngrok` or a local server, see https://github.com/linuxkit/linuxkit/blob/master/docs/platform-packet.md#boot

**Comment by cdbattags** [comment]
Any update on this for ~2021? @SvenDowideit do you have any more information about how we might go about using RacherOS's cloud-init for linuxkit images?
